Add rust overlay and update shell dependencies

markoburcul opened the pull request:
It was requested to use newer version of Rust, so I added the https://github.com/oxalica/rust-overlay to have the most recent rust toolchain. Some of the dependencies needed to compile everything (`make all`) within shell weren't there so I had to add them. Also, within the Makefile the installation of `wasm-pack` using script downloaded from the internet was replaced with installation using cargo - same result but more safe.

vinhtc27 commented:
Hey! I’ve made some changes related to the `rln-wasm` module in this [PR](https://github.com/vacp2p/zerokit/pull/296), which now requires installing `wasm-bindgen-cli` and `binaryen`. That PR is currently under review for merging into `master`.

This PR may need updates to stay in sync, so it's best to wait until the other PR is merged before reviewing.

markoburcul commented:
> This PR may need updates to stay in sync, so it's best to wait until the other PR is merged before reviewing.

Sure, just ping me when that PR is merged :)

Ben-PH commented:
`cargo-make` is needed for some build work flows, and benchmarking plots benefits from `gnuplot`.

If not added to this PR, it can roll up in one I'm currently working on as a "trivial addition of opportunity".

markoburcul commented:
> `cargo-make` is needed for some build work flows, and benchmarking plots benefits from `gnuplot`.
> 
> If not added to this PR, it can roll up in one I'm currently working on as a "trivial addition of opportunity".

I've added both to the nix shell

jakubgs commented on flake.nix:
```diff
@@ -29,9 +34,21 @@
+            rust-bin.stable.latest.default
           ];
+          # Shared library liblzma.so.5 used by wasm-pack
+          shellHook = ''
+            xz_lib=$(nix-store -q --references $(which xz) | grep xz)
+            export LD_LIBRARY_PATH=$xz_lib/lib:$LD_LIBRARY_PATH
+          '';
```
I'm not sure this is necessary, `xz` should already include the `lib` in the `env`:
```
 > nix develop
warning: Git tree '/home/jakubgs/work/zerokit' is dirty

[jakubgs@caspair:~/work/zerokit]$ echo $NIX_LDFLAGS_FOR_TARGET
-L/nix/store/bmzls7bipx954xsnbgshh8kbi488zy5m-xz-5.6.3/lib -L/nix/store/3sal5rv8apvwmbki96b11fb2sl65d6hc-rust-default-1.86.0/lib

[jakubgs@caspair:~/work/zerokit]$ echo $CMAKE_LIBRARY_PATH
/nix/store/bmzls7bipx954xsnbgshh8kbi488zy5m-xz-5.6.3/lib:/nix/store/3sal5rv8apvwmbki96b11fb2sl65d6hc-rust-default-1.86.0/lib
```

jakubgs replied:
Also, if that's not enough this can be simpler:
```suggestion
          shellHook = ''
            export LD_LIBRARY_PATH="${pkgs.xz.out}/lib:$LD_LIBRARY_PATH"
          '';
```

jakubgs commented on flake.nix:
```diff
@@ -29,9 +34,21 @@
         default = pkgs.mkShell {
-          inputsFrom = [
-            packages.${system}.default
+          buildInputs = with pkgs; [
```
I still think `inputsFrom` should be used unless not possible.

Ben-PH replied:
This was largely coming from the lengthy build time for llvm, I think, which isn't necessary for a development environment (for the most part).

markoburcul commented:
> > Sure, just ping me when that PR is merged :)
> 
> @markoburcul Hey, I just merged the other [PR](https://github.com/vacp2p/zerokit/pull/296). It should now include all the necessary dependencies. Could you update this PR to sync with the latest changes? Thanks!

I've rebased and added an option if there is someone using Nixos when installing deps via make to not run apt-get since the dependencies are provisioned in the nix shell. Also added the `nix develop` option to the docs. The `make all` runs smoothly.

jakubgs commented on flake.nix:
```diff
@@ -29,9 +34,22 @@
+            gnuplot
+            rustup
+            xz
+            wasm-pack
```
That's easy, `nixpkgs` already contains `wasm-bindgen-cli`:
https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/wasm-bindgen-cli/default.nix

markoburcul replied:
@vinhtc27 I see two options:
1. We strictly install `wasm-bindgen` and `wasm-pack` in the Makefile. Since the `installdeps` step is always executed wether you use `make all` or `make build`, these packages will be present at the build time. Regardless of using nix shell or not.
2. We add the installation of both packages to `flake.nix` file, but then you need to specify **if** the `wasm-bindgen-cli` version **needs** to be `0.2.100` since the one we can get from nix packages is `0.2.95`. I've tested this version and the **build works**.

vinhtc27 replied:
It actually just uses `wasm-bindgen` when building the `rln-wasm` module with the `multithread` feature flag inside the `/rln-wasm` directory. You can do this using:

- `cargo make build_multithread`
- `cargo make build_multithread_arkzkey`

The default build only includes the default features, just in case you didn’t know that.

I already tried using `wasm-bindgen-cli 0.2.95` to build it locallly, but it doesn't work because it requires the Rust Wasm file schema generated by version `0.2.100`, which is what I'm using in the `Cargo.toml` file of the `rln-wasm` module.


<img width="682" alt="pr comment" src="https://github.com/user-attachments/assets/39d5693b-a597-4266-9309-4e85e524bc1f" />

So I think we should choose the **1st option**, the same for `wasm-pack 0.13.1`, which requires Node.js version >= 22.

vinhtc27 replied:
So we just need to remove `wasm-pack` from the flake.nix file, and it's all good, right?

markoburcul replied:
@vinhtc27 Yes, I already did that

vinhtc27 commented:
LGTM
